pmlc/conversion/linalg_to_pxa/utils.cc

```diff
@@ -0,0 +1,77 @@
+// Copyright 2021, Intel Corporation
+
+#include "pmlc/conversion/linalg_to_pxa/pass_detail.h"
+
+namespace pmlc::conversion::linalg_to_pxa {
+
+using namespace mlir;         // NOLINT
+using namespace mlir::linalg; // NOLINT
+
+LinalgToPXATypeConverter::LinalgToPXATypeConverter() {
+  addConversion([](FunctionType type) { return type; });
+  addConversion([](FloatType type) { return type; });
+  addConversion([](IntegerType type) { return type; });
+  addConversion([](IndexType type) { return type; });
+  addConversion([](MemRefType type) { return type; });
+  addConversion([this](RankedTensorType type) {
+    Type elementType = type.getElementType();
+    Type newType = convertType(elementType);
+    assert(newType && "could not convert type");
+    return MemRefType::get(type.getShape(), newType);
+  });
+}
+
+// Generate a linalg.generic. This function uses bodyBuilder to create the
+// different generic op bodies.
+GenericOp createGenericOp(OpBuilder &builder, Operation *locationOp,
+                          TypeRange outputTypes, ValueRange inputs,
+                          ValueRange outputs, unsigned numIdxs,
+                          ArrayRef<AffineMap> maps,
+                          ArrayRef<StringRef> iterTypes,
+                          GenericOpBodyBuilder bodyBuilder) {
+  builder.setInsertionPoint(locationOp);
+  SmallVector<Value, 1> inits;
+
+  // Some original ops use outputs as operands, and some ops return outputs
+  if (outputs.empty()) {
+    // For the ops only return outputs, we have to initialize a tensor for its
+    // output.
+    for (auto outputType : outputTypes) {
+      auto shapedType = outputType.cast<ShapedType>();
+      auto outputShape = shapedType.getShape();
+      auto elemType = shapedType.getElementType();
+      auto initOp = builder.create<InitTensorOp>(builder.getUnknownLoc(),
```

Comment:
We should try to avoid using the unknown location. If you pass in the original op, we should use its location.

Reply:
Fixed.